lib/http_aws_sigv4.c

```diff
@@ -300,6 +300,76 @@ static char *parse_content_sha_hdr(struct Curl_easy *data,
   return value;
 }
 
+static CURLcode calc_payload_hash(struct Curl_easy *data,
+                                  unsigned char *sha_hash,
+                                  char *sha_hex, size_t sha_hex_size)
+{
+  const char *post_data = data->set.postfields;
+  size_t post_data_len = 0;
+  CURLcode ret = CURLE_OUT_OF_MEMORY;
+
+  if(post_data) {
+    if(data->set.postfieldsize < 0)
+      post_data_len = strlen(post_data);
+    else
+      post_data_len = (size_t)data->set.postfieldsize;
+  }
+  ret = Curl_sha256it(sha_hash, (const unsigned char *) post_data,
+                      post_data_len);
+  if(ret)
+    goto fail;
+
+  sha256_to_hex(sha_hex, sha_hash, sha_hex_size);
+fail:
+  return ret;
+}
+
+#define S3_UNSIGNED_PAYLOAD "UNSIGNED-PAYLOAD"
+
+static CURLcode calc_s3_payload_hash(struct Curl_easy *data,
+                                     Curl_HttpReq httpreq, char *provider1,
+                                     unsigned char *sha_hash,
+                                     char *sha_hex, size_t sha_hex_size)
+{
+  char *header = NULL;
+  struct curl_slist *head = NULL;
+  bool empty_method = (httpreq == HTTPREQ_GET || httpreq == HTTPREQ_HEAD);
+  /* The request method or filesize indicate no request payload */
+  bool empty_payload = (empty_method || data->set.filesize == 0);
+  /* The POST payload is in memory */
+  bool post_payload = (httpreq == HTTPREQ_POST && data->set.postfields);
+  CURLcode ret = CURLE_OUT_OF_MEMORY;
+
+  if(empty_payload || post_payload) {
+    /* Calculate a real hash when we know the request payload */
+    ret = calc_payload_hash(data, sha_hash, sha_hex, sha_hex_size);
+    if(ret)
+      goto fail;
+    ret = CURLE_OUT_OF_MEMORY;
+  }
+  else {
+    /* Fall back to s3's UNSIGNED-PAYLOAD */
+    size_t len = sizeof(S3_UNSIGNED_PAYLOAD) - 1;
+    DEBUGASSERT(len < sha_hex_size); /* 16 < 65 */
+    memcpy(sha_hex, S3_UNSIGNED_PAYLOAD, len);
+    sha_hex[len] = 0;
+  }
+
+  /* add the required content-sha256 header */
+  header = curl_maprintf("x-%s-content-sha256: %s", provider1, sha_hex);
+  if(!header)
+    goto fail;
+  head = Curl_slist_append_nodup(data->set.headers, header);
+  if(!head) {
+    free(header);
+    goto fail;
+  }
+  data->set.headers = head;
```

Comment:
the final test case 1964 doesn't add any headers to the request. in that case only, valgrind complains that this memory is leaked. is there a safe way to do this here?

below, `Curl_output_aws_sigv4()` adds a `x-amz-date` header at the end of `auth_headers`:
```c
  auth_headers = curl_maprintf("Authorization: %s4-HMAC-SHA256 "
                               "Credential=%s/%s, "
                               "SignedHeaders=%s, "
                               "Signature=%s\r\n"
                               "%s\r\n",
                               provider0,
                               user,
                               credential_scope,
                               Curl_dyn_ptr(&signed_headers),
                               sha_hex,
                               date_header);
```
is that the correct way to add `x-amz-content-sha256`?

Reply:
i added a commit that does this. the leak is gone and the tests still pass